### ISSUE TITLE
sister to groupBy

### ISSUE THREAD

**Issue by fyyyyy** [ramda/ramda#931]
From remote servers we often get arrays of objects. These objects typically contain id fields which are unique. What i would like to do is convert the array of objects in to a hashmap with the object's ids so they are quickly accessible by id.

Currenty i use groupBy, however it creates a hashmap with an array storage that wraps the object. I would like to return the object directly under the specified key.

```js

var tracks = [ {id: 1, name: "track1"}, {id: 3, name.... ]
R.indexBy('id')(tracks) //  => {1: {id: 1, name: "track1"}, 3: {id: 3, name.... }}

**Comment by fyyyyy** [comment]
This is the points free way i guess
 R.mapObj(R.head)(R.groupBy(R.get('id))(tracks));

**Comment by kedashoe** [comment]
Soon you will be able to use [into](https://github.com/ramda/ramda/blob/master/src/into.js), then you can write something like

`var indexed = into({}, tuple(get('id'), identity), tracks);`

**Comment by kedashoe** [comment]
Oops.. maybe ramda does not have a tuple-like function? I find creating a `double` is often quite handy for just this sort of situation.. so maybe `createMapEntry` with a `pipe`?

**Comment by buzzdecafe** [comment]
would `tuple` be `keyFn -> valFn -> {k: v}`  ? or `keyFn -> valFn -> [k, v]`?

**Comment by kedashoe** [comment]
> would tuple be keyFn -> valFn -> {k: v} ? or keyFn -> valFn -> [k, v]?

Maybe `tuple` does not fit ramda since it does not convey size, but in general a tuple is an ordered list, so the latter signature I think? Maybe `couple`? http://en.wikipedia.org/wiki/Tuple#Names_for_tuples_of_specific_lengths

**Comment by buzzdecafe** [comment]
you could `nTuple(n:int, [vals])`, but how is that any better than just `[vals]`?

**Comment by kedashoe** [comment]
ya I guess you could just use `Array`?

**Comment by CrossEye** [comment]
If you want it entirely points-free, and to accept `'id'` as a parameter, this might work:

    var indexBy = R.curry(R.pipe(R.useWith(R.groupBy, R.prop, R.identity), R.mapObj(R.head)));
    indexBy('id', tracks);

(I'm bothered by that explicit `curry` call.  We've let #737 linger, and I think we really need to figure it out.  But there's another issue with `compose`/`pipe` we also want to address before 1.0: their variadic nature.  So I think we're going to have another pass at that.)

This function does not seem like a great fit for Ramda itself unless we get more of a clamoring for it.  The trouble is that it makes an assumption about the data that's difficult to document, annoying to test when running, and without an obviously correct fall-back if the assumption fails: the assumption that there is only one item in the list with a given `id`.

**Comment by fyyyyy** [comment]
@CrossEye nice, thx

**Comment by benperez** [comment]
I was going to open an issue suggesting an `R.indexBy` when I decided to search for any previously existing issues. I often find myself wanting this function from ramda. 

@CrossEye I'm not convinced that 
>  the assumption that there is only one item in the list with a given id

Is reason enough to exclude this function from the library. I see this problem as the equivalent of having duplicate keys input to `R.toPairs`, where we simply take the value of the last occurrence of that key.
```javascript
> R.fromPairs([['foo', 'bar'], ['foo', 'baz']])
{ foo: 'baz' }
```

**Comment by CrossEye** [comment]
@benperez: Obviously these are close parallels.  I would still argue for a difference between the value of the first element of a pair and an arbitrary property of an object, but I certainly recognize the similarities, and I suppose we could simply document

**Comment by benperez** [comment]
I think it might be more useful if this function takes a function as it's first argument instead of a string representing a property, that way you could "index" by some computed value.

As far as the name, I think `R.keyBy` and `R.indexBy` are good options. I'll submit a PR in the next couple of days.

**Comment by buzzdecafe** [comment]
closing without prejudice for inactivity

**Comment by olsonpm** [comment]
FWIW I just now found this function missing from ramda.  Here's my use-case:

The best pattern I've found for holding data is an array of objects and to generate the corresponding maps as consumers need them.  This array of objects will almost always have one to many id fields - but consumers may prefer one to another given their different contexts.

I currently work for an insurance company and I've worked heavily with a list of coverages.  Each coverage has a unique, confusing code.  I therefore assign each coverage an abbreviation as well as a friendly string (abbreviations aren't memory friendly).  My data structure ends up like:

```js
const coverages = [
  {
    code: 'confusingUniqueBI'
    , abbr: 'BI'
    , friendly: 'Bodily Injury'
  }, {
    code: 'confusingUniquePD'
    , abbr: 'PD'
    , friendly: 'Property Damage'
  }
]
```

This structure is centralized and consumed by multiple projects.  It is in fact an array of coverages, and a dictionary would be ill-fitted since there is no one key that makes sense for every consumer.  Each consumer relies on `lodash.keyBy` to generate their own dictionary.

The above pattern has worked extremely well for me, and I feel is a perfect use-case for `keyBy`
